src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpCodeVisitor.cs

```diff
@@ -576,25 +556,38 @@ private bool ShouldGenerateInstrumentationForExpressions()
                    Context.ExpressionRenderingMode == ExpressionRenderingMode.WriteToOutput;
         }
 
-        private CSharpCodeWriter RenderPreWriteStart()
+        private void RenderStartWriteLiteral(string text)
         {
-            return RenderPreWriteStart(Writer, Context);
-        }
+            var charactersRendered = 0;
 
-        public static CSharpCodeWriter RenderPreWriteStart(CSharpCodeWriter writer, CodeGeneratorContext context)
-        {
-            if (!string.IsNullOrEmpty(context.TargetWriterName))
+            // Render the string in pieces to avoid Roslyn OOM exceptions at compile time:
+            // https://github.com/aspnet/External/issues/54
+            while (charactersRendered < text.Length)
             {
-                writer.WriteStartMethodInvocation(context.Host.GeneratedClassContext.WriteLiteralToMethodName)
-                      .Write(context.TargetWriterName)
-                      .WriteParameterSeparator();
-            }
-            else
-            {
-                writer.WriteStartMethodInvocation(context.Host.GeneratedClassContext.WriteLiteralMethodName);
-            }
+                if (Context.ExpressionRenderingMode == ExpressionRenderingMode.WriteToOutput)
+                {
+                    if (!string.IsNullOrEmpty(Context.TargetWriterName))
+                    {
+                        Writer.WriteStartMethodInvocation(Context.Host.GeneratedClassContext.WriteLiteralToMethodName)
+                              .Write(Context.TargetWriterName)
+                              .WriteParameterSeparator();
+                    }
+                    else
+                    {
+                        Writer.WriteStartMethodInvocation(Context.Host.GeneratedClassContext.WriteLiteralMethodName);
+                    }
+                }
 
-            return writer;
+                var charactersToSubstring = Math.Min(MaxStringLiteralLength, text.Length - charactersRendered);
+                var textToRender = text.Substring(charactersRendered, charactersToSubstring);
```

Comment:
There's a difference between deferring perf work and adding a useless `Substring()` to the mainline code path. Keep the `text.Length < MaxStringLiteralLength` case. That's the code path that'll be used 99% of the time this method is called.